feat(blog): explain with basic scan types (#1535)
questdb/questdb.io: blog/2023-04-20-explain-basic-scans.md

---
title: "EXPLAIN scan types"
author: Bolek Ziobrowski
author_title: QuestDB Team
author_url: https://github.com/bziobrowski
author_image_url: https://avatars.githubusercontent.com/bziobrowski
description:
  A tour of scan types available in QuestDB.
image: /img/blog/2023-04-20/banner.jpg
tags: [explain, SQL, execution plan, table scan, index scan]
slug: explain-sql-index-table-scan
---

<!--truncate-->

import Banner from "@theme/Banner"

<Banner
alt="Image of a hard disk drive."
height={433}
src="/img/blog/2023-04-20/banner.jpg"
width={650}
>
Photo by <a href="https://unsplash.com/@benjaminlehman">Benjamin Lehman</a> via{" "}
<a href="https://unsplash.com">Unsplash</a>
</Banner>

Welcome to another post on sql performance tuning !
This time we'll explore the various scan types (a.k.a. access methods) QuestDB supports.
It is necessary to understand them before tackling more complex queries. 

> What are scan types  ? 

Reviewer: "scan type" or "access mode" does not seem to be a mainstream keywords. 
I googled some terms like database scan but it does not convey the same thing. Is there a better term widely used that we could use here?

If not, we could use a cluster of strong keywords to describe the concept in the header.

We need a good level twl header here before jumping into each type.

Scan type is an algorithm used to find and access data. 
Imagine binary search on sorted data, but with partitions, indexes and column data files instead of arrays.   

> There are two basic scan types in QuestDB, table and index scan.

Author: If we use standard terms then interval scan is like a sub-type of full table scan but if we use ours then we can organize it like this : 

- Table scans 
  - Frame forward/backward scan 
  - Interval forward/backward scan 
- Index scans 
  - Index forward/backward scan in index order 
  - Index forward/backward scan in table order 

What do you think ?

> There are two basic scan types in QuestDB, table and index scan.

Reviewer: We may not want to keep the headers exactly the same but I think overall this structure works better:
- Table scans
   - Frame (forward/backward) scan
   [diagram indicating both directions]
      - text and examples
   - Interval (forward/backward) scan
      [diagram indicating both directions]
      - text and examples
- Index scans
   - Index (forward/backward) scan in index order
        [diagram indicating both directions]
      - text and examples
  - Index (forward/backward) scan in table order
     [diagram indicating both directions]
      - text and examples

> There are two basic scan types in QuestDB, table and index scan.

Reviewer: Apologies - the structure I made in [e57622b](https://github.com/questdb/questdb.io/pull/1535/commits/e57622b378fe678819e4c9fc97d3d64bb3cbd6de) is what I think would work - I did keep the single key part!

The difference is that the former touches table data directly, while latter goes through index first. 
Both occur in a number of variants so let's go through the list and explore them in depth.

All examples below use tables available in the [QuestDB demo instance](https://demo.questdb.io/).
Schema: 

```questdb-sql
CREATE TABLE trades (
 symbol SYMBOL,
 side SYMBOL,
 price DOUBLE,
 amount DOUBLE,
 timestamp TIMESTAMP
) timestamp (timestamp) PARTITION BY DAY WAL;
```   

> 

Author: I'm mixing terms here a bit because the common term is "Table scan" or "Full table scan" while explain's output shows "Frame forward scan" . Do you think it's ok as is or should I change it

Author: Maybe a bit of explanation would do ? 
The problem is that our terms don't entirely match the standard ones and EXPAIN's output is custom . 
Normally "Full table scan" is done from start to end of table because data is not ordered and direction doesn't matter . Qdb tables that have designated timestamp are ordered by it so direction matters a lot . 

Interval scan : 
- conceptually is a sub-type of full table scan 
- can be done in forward or backward direction
- is specific to qdb . 

Index scans are more straightforward, expect that our indexes aren't the usual b-trees but hash-map-like .

Reviewer: UUnderstood.
But is this clear in the article?

> expect that our indexes aren't the usual b-trees but hash-map-like

<Banner
alt="Table forward scan."
height={433}
src="/img/blog/2023-04-20/forward_scan.svg"
width={650}
/>

#### Table forward scan

Scans all table rows starting at first row of the oldest partition and ending at last row of the latest partition.
Differentiating between forward and backward scan makes sense only for tables with designated timestamp because they store data in that timestamp order.
For tables without designated timestamp scan direction doesn't make a difference because there's no predictable order to data.
Let's check a simple select: 

```questdb-sql
explain select * from trades 
order by timestamp
```

 | QUERY PLAN                                            |
|-------------------------------------------------------|
| DataFrame                                             |
| &nbsp;&nbsp;&nbsp;&nbsp;Row forward scan              |
| &nbsp;&nbsp;&nbsp;&nbsp;Frame forward scan on: trades |

As expected, optimizer recognized that `order by timestamp` can be implemented simply by forward scanning the table, no sorting required.
In `EXPLAIN` output a Table forward scan is represented on table level - `Frame forward scan` and partition level - `Row forward scan`.

> 

Author: @amyshwang I included "backward" version for table scan only because I felt it would be too repetitive . 
Let me know if you think we should include it for all scan types .

Reviewer: I think it is important to be logical so:

- (full) table scan
    - Frame forward scan
    - Frame backward scan

This is very clear.
What we should consider more is to add more descriptive headers because Frame forward/backward scan is QuestDB terms.

Author: I meant that diagram-wise . Do you think I should add separate diagram for interval backward scan, index backward scan, ... ?

Reviewer: I guess we could do a diagram for each scan type with both directions added to the diagrams? For example:

![Screenshot 2023-04-13 at 12 46 03](https://user-images.githubusercontent.com/107401731/231656391-0c22ef48-5cf1-419e-a9e1-954e656365fd.png)

Reviewer: I moved the existing diagrams to places that I think suitable and they can be combined to one diagram for both directions.

Author: For table scans it should be fine but wouldn't we need to change indices on index scan diagrams to show rows being read in backward direction ?

Reviewer: For table order and index order scan, I don't think the arrows for forward scan makes sense because (if I understand it correctly) you have numbered the scan sequence.

We could use the diagram, remove the arrow and add:

Index forward scan -multiple keys, table order]: scanning from (1) to (6) sequentially
Index forward scan -multiple keys, table order]: scanning from (6) to (1) in decreasing order

Author: The arrow is not necessary, it just shows the scan direction on partition level .

```questdb-sql
explain select * from trades 
where timestamp in '2023-01-20'
```

 | QUERY PLAN                                                                                 |
|--------------------------------------------------------------------------------------------|
| DataFrame                                                                                  |
| &nbsp;&nbsp;&nbsp;&nbsp;Row forward scan                                                   |
| &nbsp;&nbsp;&nbsp;&nbsp;Interval forward scan on: trades                                   |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;intervals: [static=[1674172800000000,1674259199999999] |

Plan above shows that optimizer reduced scanning to single interval equal to 2023-01-20 day partition.  
Engine might even detect conflicting conditions and not run any scan at all, e.g.

```questdb-sql
explain select * from trades 
where timestamp in '2023-01-20' 
and timestamp < '2022-01-01'
```
produces :

| QUERY PLAN     |
|----------------|
| Empty table    |

If predicate is too complex (especially if designated timestamp is used as function argument), engine will fall back to default table scan with filter, e.g. 

```questdb-sql

> explain select * from trades 

Reviewer: Let's use the correct syntax and capitalisation?

where dateadd('m', 1, timestamp) in '2023-01-20'
```

| QUERY PLAN                                                                          |
|-------------------------------------------------------------------------------------|
| Async Filter                                                                        |
| &nbsp;&nbsp;filter: dateadd('m',1,timestamp) in [1674172800000000,1674259199999999] |
| &nbsp;&nbsp;workers: 24                                                             |
| &nbsp;&nbsp;&nbsp;&nbsp;DataFrame                                                   |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Row forward scan                    |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Frame forward scan on: trades       |

Rewriting predicate as `timestamp between '2023-01-19T23:59:00.000000Z' and '2023-01-20T23:59:00.000000Z'` makes query use interval filter.  
Lesson - keep designated timestamp predicates clean and simple!

#### Interval backward scan

> Same as forward scan but runs from last row of last interval to first row of first interval.   

Reviewer: This applies elsewhere too - I'd use complete sentences.

```questdb-sql
EXPLAIN
SELECT * FROM trades
WHERE timestamp in '2023-01-20'
ORDER BY timestamp DESC
```

| QUERY PLAN                                                                                 |
| ------------------------------------------------------------------------------------------ |
| DataFrame                                                                                  |
| &nbsp;&nbsp;&nbsp;&nbsp;Row backward scan                                                  |
| &nbsp;&nbsp;&nbsp;&nbsp;Interval backward scan on: trades                                  |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;intervals: [static=[1669852800000000,1669939199999999] |

## Index scan

Index scan first reads row id data associated with one or more index keys and then respective data from the table. See documentation for more information on indexes. The `trades` table doesn't have any index, so let's switch to another demo table, `pos`, with the following schema:

```questdb-sql
CREATE TABLE pos (
 time TIMESTAMP,
 id SYMBOL INDEX,
 lat DOUBLE,
 lon DOUBLE,
 geo6 GEOHASH(6c),
 geo12 GEOHASH(12c)
) timestamp (time) PARTITION BY DAY;
```

> 

Reviewer: Is it right to insert a diagram here? It should be in "Index scan with a single key"?

| DeferredSingleSymbolFilterDataFrame                                                        |
| &nbsp;&nbsp;&nbsp;&nbsp;Index backward scan on: id deferred: true                          |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;filter: id='X'                                         |
| &nbsp;&nbsp;&nbsp;&nbsp;Interval forward scan on: pos                                      |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;intervals: [static=[1675209600000000,1675295999999999] |

As you can see, not only is the potential sort replaced by a backward scan, but
the scanning was reduced to a single timestamp interval.

### Index scan with multiple keys

Things get more interesting when there's more than one index key to scan. We can
scan row ids in the following ways:

- `Table-order` - scans the minimum row id available from all per-key row ids
  until there's none left.  
- `Index-order` - first reads all row ids associated with key k1, then k2, etc.

Both have pros and cons, so let's look at a simple example.

#### Table order

<Banner
  alt="Index forward scan in table order."
  height={433}
  src="/img/blog/2023-04-20/index_scan_table_order.svg"
  width={650}

> />

Reviewer: The arrow only shows a forward scan. I wonder if my previous suggestion [here](https://github.com/questdb/questdb.io/pull/1535/files/57ea3e9071d825d181ee4297adca22c7993694c6#r1166356099) still applies?
> 
> We could use the diagram, remove the arrow and add:
> 
> Index forward scan -multiple keys, table order]: scanning from (1) to (6) sequentially
> Index forward scan -multiple keys, table order]: scanning from (6) to (1) in decreasing order

And the same for the index order diagram, where the arrow does not make sense.

Author: I removed arrows but netlify still shows old diagrams.

Reviewer: Probably due to the cache.
I am not sure if it is obvious that the number in parenthesis indicates the order though. It took me a while to figure that out and I am still not sure if that's correct. Could you add some text to explain it?

| FilterOnValues                                                                           |
| &nbsp;&nbsp;&nbsp;&nbsp;Table-order scan                                                 |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Index forward scan on: id deferred: true |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;filter: id='X'               |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Index forward scan on: id deferred: true |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;filter: id='Y'               |
| &nbsp;&nbsp;&nbsp;&nbsp;Frame forward scan on: pos                                       |

#### Index order

Index scan with multiple values in index order scans table rows using row ids associated with the first index key value, followed by the second and the third. The scan continues until reaching the last index value. It might be used to avoid sorting at the price of potentially more random memory and disk accesses.

<Banner
  alt="Index forward scan in index order."
  height={433}
  src="/img/blog/2023-04-20/index_forward_scan_index_order.svg"
  width={650}

> />

Reviewer: I forgot to move this to L313

```questdb-sql
EXPLAIN SELECT * FROM pos
WHERE id in ('X', 'Y')
AND time in '2023-02-01'
ORDER BY id,time DESC
```

| QUERY PLAN                                                                                 |
| ------------------------------------------------------------------------------------------ |
| FilterOnValues symbolOrder: asc                                                            |
| &nbsp;&nbsp;&nbsp;&nbsp;Cursor-order scan                                                  |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Index backward scan on: id deferred: true  |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;filter: id='X'                 |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Index backward scan on: id deferred: true  |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;filter: id='Y'                 |
| &nbsp;&nbsp;&nbsp;&nbsp;Interval forward scan on: pos                                      |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;intervals: [static=[1675209600000000,1675295999999999] |

> Note that while technically we're doing an interval forward scan, it doesn't really matter because there's just one interval to scan. While this makes sense with a single partition, it doesn't extend to the case with multiple partitions. That's because, on a higher level, the SQL engine iterates over partitions, and in doing so it mixes the order.

Reviewer: I don't understand this because the plan says "Index backward scan"...Why is it technically a forward scan? Because it's sequential to the key?

Reviewer: You did say:
> In the EXPLAIN output, a table forward scan is represented on the table level as Frame forward scan, and on the partition level, Row forward scan.

But it's under Table scan and by the time I am at Index scan, I don't remember that piece of info anymore @@ 

Maybe moving that line to the intro, before "Table scan" and after:

> QuestDB supports two main scan types, table and index scan. The difference is that the former touches table data directly, while the latter goes through the index first. Both types include sub-types and variants.

So in the intro we talk about:
- scan types
- scan levels

Reviewer: I think this paragraph has not been updated and is still a bit confusing.

---

> title: "EXPLAIN scan types"

Reviewer: We will need to change the title and slug

time-partitioned tables, often with tens or hundreds of partitions. If we were
to follow PostgreSQL's approach of showing all partitions, for instance:

```
EXPLAIN
SELECT *
FROM trades
WHERE timestamp < now()
```

we will get:

| QUERY PLAN                                                                |
| ------------------------------------------------------------------------- |
| Append (cost=0.00..176.00 rows=2720 width=12)                             |
| -> Seq Scan on trades_y2006 trades_1 (cost=0.00..40.60 rows=680 width=12) |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Filter: (ts < now())            |
| -> Seq Scan on trades_y2021 trades_2 (cost=0.00..40.60 rows=680 width=12) |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Filter: (ts < now())            |
| ...                                                                       |

This output would take too much space.

Therefore, QuestDB's EXPLAIN command shows:

- The direction of Data Frame (a piece or the whole partition) iteration, e.g.
  `Frame forward scan`
- The direction of Row iteration within a Data Frame, e.g. `Row forward scan`
- A list of scan boundaries in `Interval scan`, e.g.

>   `intervals: [("1970-01-01T00:00:00.000000Z","1970-01-01T23:59:59.999999Z")]`

Reviewer: I feel this important info should be in the intro as it is not just for Table scan?

Author: I put it here because it makes them most sense with an example .

| &nbsp;&nbsp;&nbsp;&nbsp;Async JIT Filter                                      |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;limit: 1                                  |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;filter: dropoff_location_id=110           |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;workers: 24                               |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;DataFrame                     |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Row backward scan             |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Frame backward scan on: trips |

As expected, the query does a backward scan (`Frame backward scan`) and stops on
the first matching row ( `limit: 1` under `Async JIT Filter` ).  
Remember that the approach above only makes sense if the data is close to the
end of the table; otherwise, it might slow things down.

## Using `EXPLAIN` to optimize your SQL queries

We have introduced the basic scan types available in QuestDB:

- Frame scan
- Interval scan
- Index scan (in table and index order)

We have also demonstrated the available optimizations for tables with the
designated timestamp.

> We hope you enjoy this tutorial.

Reviewer: Could we add an ending highlighting the benefit of using EXPLAIN to refine SQL queries?

Author: Added in latest commit .

| ---------------------------------------------------------------------------------------- |
| FilterOnValues                                                                           |
| &nbsp;&nbsp;&nbsp;&nbsp;Table-order scan                                                 |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Index forward scan on: id deferred: true |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;filter: id='X'               |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Index forward scan on: id deferred: true |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;filter: id='Y'               |
| &nbsp;&nbsp;&nbsp;&nbsp;Frame forward scan on: pos                                       |

#### Index order

<Banner
  alt="Index scan in index order."
  height={433}
  src="/img/blog/2023-04-20/index_scan_index_order.svg"
  width={650}
>
  Index forward scan in index order: scanning from (1) to (6) 
  Index backward scan in index order: scanning from (6) to (1)
</Banner>

Index scans with multiple keys in index order scan table rows using row ids
associated with the first index key, followed by those associated with the

> second index key, and so on.

Reviewer: The description and the diagram seem to be contradictory to each other.

<img width="674" alt="Screenshot 2023-04-21 at 14 31 49" src="https://user-images.githubusercontent.com/107401731/233558572-1bda42b3-b8f5-446e-9a86-8f13e68fa9f0.png">

>Index scans with multiple keys in index order scan table rows using row ids
associated with the first index key, followed by those associated with the
second index key, and so on.

The diagram shows scanning table rows using row ids
associated with the first index key in the first partition, followed by the table row ids with the second index key in the same partition. The scan then continues to the second partition.

Author: I guess the explanation on two levels of iteration/scanning isn't clear enough . 
Indexes are partitioned  and scanning on the top level happens on partitions or intervals and for each such 'Frame' runs a  row or index scan .

Reviewer: Might it be better to update this paragraph too?
```suggestion
Multiple-key-index-order scans iterate one partition after another,  scanning table rows using row ids associated with the first index key, followed by those associated with the second index key, and so on.
```

selecting a single/a few columns from a wide table might read just a few percent
of the table's data.

> Frame forward scans start at the first row of the oldest partition (one with lowest timestamp partition key value) 

Reviewer: the one with the lowest timestamp, and the one with the highest timestamp in L86

- The direction of Data Frame (a piece or the whole partition) iteration, e.g.
  `Frame forward scan`
- The direction of Row iteration within a Data Frame, e.g. `Row forward scan`

> - A list of scan boundaries in `Interval scan`, e.g.\

Reviewer: ```suggestion
- A list of scan boundaries in `Interval scan`, e.g.
```

| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Row forward scan                    |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Frame forward scan on: trades       |

> Rewriting the `WHERE` condition to\
> `timestamp BETWEEN '2023-01-19T23:59:00.000000Z' AND '2023-01-20T23:59:00.000000Z'`\

Reviewer: ```suggestion
Rewriting the `WHERE` condition to
`timestamp BETWEEN '2023-01-19T23:59:00.000000Z' AND '2023-01-20T23:59:00.000000Z'`
```

Things get more interesting when there's more than one index key to scan. As
shown in the index scan diagram above, indexes are partitioned. When QuestDB
iterates over a partition or interval list and conducts an index scan for each

> partition or interval, the output still in timestamp order.

Reviewer: ```suggestion
partition or interval, with the output still in timestamp order.
```
